src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs

```diff
@@ -199,6 +226,18 @@ internal partial class DiagnosticIncrementalAnalyzer
                         return new ProjectAnalysisData(project.Id, VersionStamp.Default, existingData.Result, results);
                     }
 
+                    // Reduce the state sets to analyze based on individual full solution analysis values
+                    // for compiler diagnostics and analyzers.
+                    if (!compilerFullAnalysisEnabled)
+                    {
+                        Debug.Assert(analyzersFullAnalysisEnabled);
+                        stateSets = stateSets.Where(s => !s.Analyzer.IsCompilerAnalyzer());
```

Comment:
this implies taht state-sets is an IENumerable.  which means each time we process it, it will be enumerated and executed lazily.  strongly consider making this a realized list/array.